File: spring-data-requery/src/main/java/org/springframework/data/requery/cache/cache2k/Cache2kEntityCache.java
package org.springframework.data.requery.cache.cache2k;

import io.requery.EntityCache;
import io.requery.sql.EntityDataStore;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.cache2k.Cache;
import org.cache2k.Cache2kBuilder;
import org.cache2k.configuration.Cache2kConfiguration;

import javax.annotation.Nonnull;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Cache2kEntityCache
 *
 * @author debop
 * @since 19. 3. 11
 */
@SuppressWarnings("unchecked")
@Slf4j
public class Cache2kEntityCache implements EntityCache {

    private static final Cache2kConfiguration<Object, Object> defaultConfiguration;

    private static final Map<Class<?>, Cache<Object, Object>> cacheManagers = new ConcurrentHashMap<>();
    private static final Object syncObj = new Object();

    static {
        defaultConfiguration = Cache2kConfiguration.of(Object.class, Object.class);
        defaultConfiguration.setEntryCapacity(20000L);
        defaultConfiguration.setEternal(true);
        defaultConfiguration.setKeepDataAfterExpired(false);
        defaultConfiguration.setBoostConcurrency(true);
        defaultConfiguration.setRetryInterval(10);
        defaultConfiguration.setMaxRetryInterval(1000L);
    }

    private final Cache2kConfiguration<Object, Object> configuration;

    // For auto loading
    @Getter(AccessLevel.PROTECTED)
    @Setter
    private EntityDataStore<Object> dataStore = null;

    public Cache2kEntityCache() {
        this(defaultConfiguration);
    }

    public Cache2kEntityCache(@Nonnull final Cache2kConfiguration<Object, Object> configuration) {
        this.configuration = configuration;
    }

    private Cache getCache(Class<?> entityType) {
        return cacheManagers.computeIfAbsent(entityType, (key) -> {
            log.debug("Create cache2k cache for type {}", key.getName());

            Cache2kBuilder<Object, Object> builder = Cache2kBuilder.of(configuration).name(key.getName());

            // For auto loading
            if (dataStore != null) {
                builder.loader((id) -> dataStore.findByKey(key, id));
            }
            return builder.build();
        });
    }

    @Override
    public <T> T get(Class<T> type, Object key) {
        return (T) getCache(type).get(key);
    }

    @Override
    public <T> void put(Class<T> type, Object key, T value) {
        if (value != null) {
            getCache(type).put(key, value);
        } else {
            invalidate(type, key);
        }
    }

    @Override
    public boolean contains(Class<?> type, Object key) {
        return getCache(type).containsKey(key);
    }

    @Override
    public void invalidate(Class<?> type) {
        getCache(type).removeAll();
    }

    @Override
    public void invalidate(Class<?> type, Object key) {
        getCache(type).remove(key);
    }

    @Override
    public void clear() {
        synchronized (syncObj) {
            cacheManagers.forEach((type, cache) -> cache.clearAndClose());
            cacheManagers.clear();
        }
    }
}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/cache/Cache2kEntityCache.kt
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.kotlin.cache

import io.requery.EntityCache
import io.requery.sql.EntityDataStore
import mu.KLogging
import org.cache2k.Cache
import org.cache2k.Cache2kBuilder
import org.cache2k.configuration.Cache2kConfiguration
import java.util.concurrent.ConcurrentHashMap

/**
 * Cache2k를 이용하여 Requery entity를 cache 합니다.
 * Requery Entity는 분산환경에서의 Cache는 사용할 수 없습니다 (DTO로 만든 후 캐시해야 합니다)
 *
 * @author debop
 * @since 18. 7. 2
 */
class Cache2kEntityCache @JvmOverloads constructor(
    val configuration: Cache2kConfiguration<Any, Any?> = DEFAULT_CONFIGURATION) : EntityCache {

    companion object : KLogging() {

        private val DEFAULT_CONFIGURATION: Cache2kConfiguration<Any, Any?> =
            Cache2kConfiguration.of(Any::class.java, Any::class.java)
                .apply {
                    entryCapacity = 20000L
                    isEternal = true
                    isKeepDataAfterExpired = false
                    isBoostConcurrency = true
                    retryInterval = 10
                    maxRetryInterval = 1000L
                }

        private val cacheManager: MutableMap<Class<*>, Cache<Any, Any?>> = ConcurrentHashMap()
        private val syncObj: Any = Any()
    }

    var dataStore: EntityDataStore<Any>? = null

    val Class<*>.cache: Cache<Any, Any?>
        get() {
            return cacheManager.computeIfAbsent(this) { clazz ->
                synchronized(syncObj) {
                    logger.debug { "Create Cache2k cache for type [${this.name}]" }

                    Cache2kBuilder.of(configuration).name(clazz.name).run {
                        if(dataStore != null) {
                            loader { dataStore!!.findByKey(clazz, it) }
                        }
                        build()
                    }
                }
            }
        }

    override fun contains(type: Class<*>?, key: Any?): Boolean {
        return type?.cache?.containsKey(key) ?: false
    }

    override fun invalidate(type: Class<*>?) {
        type?.cache?.removeAll()
    }

    override fun invalidate(type: Class<*>?, key: Any?) {
        key?.let { type?.cache?.remove(key) }
    }

    override fun clear() {
        synchronized(syncObj) {
            cacheManager.forEach { _, cache ->
                cache.clearAndClose()
            }
            cacheManager.clear()
        }
    }

    override fun <T : Any> put(type: Class<T>?, key: Any?, value: T?) {
        value?.let {
            key?.let { type?.cache?.put(key, value) }
        } ?: invalidate(type, key)
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : Any> get(type: Class<T>?, key: Any?): T? {
        return key?.let {
            type?.cast(type.cache.get(key))
        }
    }

}